File: backend/scripts/seed_mxfood_to_s3.py
"""One-time script: read mxfood.duckdb and upload all tables as Parquet to S3.

Usage:
    cd backend
    uv run python scripts/seed_mxfood_to_s3.py

Required environment variables:
    OBJECT_STORE_BUCKET              — S3 bucket name (required)
    OBJECT_STORE_REGION              — AWS region (default: us-east-1)
    OBJECT_STORE_ACCESS_KEY_ID       — AWS access key
    OBJECT_STORE_SECRET_ACCESS_KEY   — AWS secret key
    OBJECT_STORE_ENDPOINT            — Custom endpoint (optional, for MinIO/R2)

Optional:
    BASE_DUCKDB_DATA_PATH            — Base directory for resolving DuckDB paths (default: ..)
"""

import os
import sys

import duckdb


def resolve_duckdb_path(file_path: str) -> str:
    base = os.environ.get("BASE_DUCKDB_DATA_PATH", "..")
    if os.path.isabs(file_path):
        return file_path
    return os.path.normpath(os.path.join(base, file_path))


def main() -> None:
    bucket = os.environ.get("OBJECT_STORE_BUCKET")
    if not bucket:
        print("ERROR: OBJECT_STORE_BUCKET environment variable is not set.", file=sys.stderr)
        sys.exit(1)

    region = os.environ.get("OBJECT_STORE_REGION", "us-east-1")
    access_key = os.environ.get("OBJECT_STORE_ACCESS_KEY_ID", "")
    secret_key = os.environ.get("OBJECT_STORE_SECRET_ACCESS_KEY", "")
    endpoint = os.environ.get("OBJECT_STORE_ENDPOINT", "")

    db_path = resolve_duckdb_path("data/mxfood.duckdb")
    print(f"Opening DuckDB: {db_path}")

    con = duckdb.connect(db_path, read_only=True)

    # Install and load httpfs for S3 support
    con.execute("INSTALL httpfs")
    con.execute("LOAD httpfs")

    # Configure S3 credentials
    con.execute(f"SET s3_region = '{region}'")
    con.execute(f"SET s3_access_key_id = '{access_key}'")
    con.execute(f"SET s3_secret_access_key = '{secret_key}'")
    if endpoint:
        con.execute(f"SET s3_endpoint = '{endpoint}'")
        con.execute("SET s3_url_style = 'path'")

    # Enumerate all user tables in the main schema
    tables_result = con.execute(
        "SELECT table_name FROM information_schema.tables "
        "WHERE table_schema = 'main' AND table_type = 'BASE TABLE' "
        "ORDER BY table_name"
    ).fetchall()

    tables = [row[0] for row in tables_result]
    if not tables:
        print("No tables found in main schema of mxfood.duckdb.", file=sys.stderr)
        sys.exit(1)

    print(f"Found {len(tables)} tables: {', '.join(tables)}")
    print()

    succeeded = []
    failed = []

    for table in tables:
        s3_path = f"s3://{bucket}/seeds/mxfood/{table}.parquet"
        print(f"  Uploading {table} → {s3_path} ...", end=" ", flush=True)
        try:
            con.execute(
                f'COPY (SELECT * FROM main."{table}") TO \'{s3_path}\' (FORMAT PARQUET)'
            )
            print("OK")
            succeeded.append(table)
        except Exception as exc:
            print(f"FAILED: {exc}")
            failed.append((table, str(exc)))

    con.close()

    print()
    print(f"Done. {len(succeeded)}/{len(tables)} tables uploaded successfully.")
    if succeeded:
        print(f"  Succeeded: {', '.join(succeeded)}")
    if failed:
        print(f"  Failed: {', '.join(t for t, _ in failed)}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: frontend/scripts/seed-mxfood-to-s3.ts
#!/usr/bin/env tsx
/**
 * One-time script: download mxfood.duckdb from GitHub releases and upload all
 * tables as Parquet to S3. These become the shared seed files copied per-company
 * at registration time:
 *   seeds/mxfood/{table}.parquet → {companyId}/csvs/tutorial/mxfood/{table}.parquet
 *
 * Usage (run from frontend/):
 *   npm run seed-mxfood
 *
 * Reads env vars from frontend/.env automatically.
 *
 * Required:
 *   OBJECT_STORE_BUCKET              S3 bucket name
 *   OBJECT_STORE_ACCESS_KEY_ID       AWS access key
 *   OBJECT_STORE_SECRET_ACCESS_KEY   AWS secret key
 *
 * Optional:
 *   OBJECT_STORE_REGION              AWS region (default: us-east-1)
 *   OBJECT_STORE_ENDPOINT            Custom endpoint for MinIO/R2 (e.g. http://minio:9000)
 *   MXFOOD_DUCKDB_URL                Override download URL (default: GitHub releases)
 */

import { config } from 'dotenv';
import { tmpdir } from 'os';
import { join } from 'path';
import { writeFile, unlink } from 'fs/promises';
import { DuckDBInstance } from '@duckdb/node-api';

config();

const BUCKET = process.env.OBJECT_STORE_BUCKET;
const REGION = process.env.OBJECT_STORE_REGION ?? 'us-east-1';
const ACCESS_KEY = process.env.OBJECT_STORE_ACCESS_KEY_ID ?? '';
const SECRET_KEY = process.env.OBJECT_STORE_SECRET_ACCESS_KEY ?? '';
const ENDPOINT = process.env.OBJECT_STORE_ENDPOINT ?? '';
const MXFOOD_URL =
  process.env.MXFOOD_DUCKDB_URL ??
  'https://github.com/minusxai/sample_datasets/releases/download/v1.0/mxfood.duckdb';

if (!BUCKET) {
  console.error('ERROR: OBJECT_STORE_BUCKET is not set. Add it to frontend/.env');
  process.exit(1);
}

async function main() {
  // Download mxfood.duckdb to a temp file
  console.log(`Downloading: ${MXFOOD_URL}`);
  const res = await fetch(MXFOOD_URL);
  if (!res.ok) throw new Error(`Download failed: ${res.status} ${res.statusText}`);
  const buffer = Buffer.from(await res.arrayBuffer());

  const tmpPath = join(tmpdir(), `mxfood-seed-${Date.now()}.duckdb`);
  await writeFile(tmpPath, buffer);
  console.log(`Downloaded ${(buffer.length / 1024 / 1024).toFixed(1)} MB → ${tmpPath}`);
  console.log(`Uploading to: s3://${BUCKET}/seeds/mxfood/\n`);

  const instance = await DuckDBInstance.create(tmpPath);
  const conn = await instance.connect();

  try {
    await conn.run('INSTALL httpfs');
    await conn.run('LOAD httpfs');
    await conn.run(`SET s3_region = '${REGION}'`);
    if (ACCESS_KEY) await conn.run(`SET s3_access_key_id = '${ACCESS_KEY}'`);
    if (SECRET_KEY) await conn.run(`SET s3_secret_access_key = '${SECRET_KEY}'`);
    if (ENDPOINT) {
      await conn.run(`SET s3_endpoint = '${ENDPOINT}'`);
      await conn.run("SET s3_url_style = 'path'");
    }

    const result = await conn.run(
      "SELECT table_name FROM information_schema.tables " +
      "WHERE table_schema = 'main' AND table_type = 'BASE TABLE' " +
      "ORDER BY table_name"
    );
    const rows = await result.getRowObjectsJS() as Array<{ table_name: string }>;
    const tables = rows.map(r => r.table_name);

    if (tables.length === 0) {
      console.error('No tables found in mxfood.duckdb');
      process.exit(1);
    }
    console.log(`Found ${tables.length} tables: ${tables.join(', ')}\n`);

    const succeeded: string[] = [];
    const failed: Array<[string, string]> = [];

    for (const table of tables) {
      const s3Path = `s3://${BUCKET}/seeds/mxfood/${table}.parquet`;
      process.stdout.write(`  ${table} → ${s3Path} ... `);
      try {
        await conn.run(`COPY (SELECT * FROM main."${table}") TO '${s3Path}' (FORMAT PARQUET)`);
        console.log('OK');
        succeeded.push(table);
      } catch (err: unknown) {
        const msg = err instanceof Error ? err.message : String(err);
        console.log(`FAILED: ${msg}`);
        failed.push([table, msg]);
      }
    }

    console.log(`\nDone. ${succeeded.length}/${tables.length} tables uploaded successfully.`);
    if (failed.length > 0) {
      console.error(`  Failed: ${failed.map(([t]) => t).join(', ')}`);
      process.exit(1);
    }
  } finally {
    conn.closeSync();
    await unlink(tmpPath).catch(() => {}); // clean up temp file
  }
}

main().catch(err => {
  console.error('Fatal:', err);
  process.exit(1);
});
